Source/quests.cpp

```diff
@@ -721,59 +745,95 @@ void ResyncQuests()
 
 void DrawQuestLog(const Surface &out)
 {
-	DrawString(out, _("Quest Log"), { GetPanelPosition(UiPanels::Quest, { 32, 44 }), { 257, 0 } }, UiFlags::AlignCenter);
+	int l = QuestLogMouseToEntry();
+	if (l >= 0) {
+		selectedEntry = l;
+	}
+	const auto x = panelInnerRect.position.x;
 	CelDrawTo(out, GetPanelPosition(UiPanels::Quest, { 0, 351 }), *pQLogCel, 1);
-	int line = qtopline;
-	for (int i = 0; i < numqlines; i++) {
-		PrintQLString(out, 32, line, _(QuestData[qlist[i]]._qlstr));
-		line += 2;
+	int y = panelInnerRect.position.y + topY;
+	for (int i = 0; i < qlistCnt; i++) {
+		if (i == firstFinishedEntry) {
+			y += act2finSpacing;
+		}
+		PrintQLString(out, x, y, _(QuestData[qlist[i]]._qlstr), i == selectedEntry, i >= firstFinishedEntry);
+		y += lineSpacing;
 	}
-	PrintQLString(out, 32, 22, _("Close Quest Log"));
 }
 
 void StartQuestlog()
 {
-	numqlines = 0;
+
+	auto sortQuestIdx = [](int a, int b) {
+		return QuestData[a].questBookOrder < QuestData[b].questBookOrder;
+	};
+
+	qlistCnt = 0;
 	for (auto &quest : Quests) {
 		if (quest._qactive == QUEST_ACTIVE && quest._qlog) {
-			qlist[numqlines] = quest._qidx;
-			numqlines++;
+			qlist[qlistCnt] = quest._qidx;
+			qlistCnt++;
 		}
 	}
-	if (numqlines > 5) {
-		qtopline = 5 - (numqlines / 2);
-	} else {
-		qtopline = 8;
+	firstFinishedEntry = qlistCnt;
+	for (auto &quest : Quests) {
+		if (quest._qactive == QUEST_DONE || quest._qactive == QUEST_HIVE_DONE) {
+			qlist[qlistCnt] = quest._qidx;
+			qlistCnt++;
+		}
+	}
+
+	std::sort(&qlist[0], &qlist[firstFinishedEntry], sortQuestIdx);
+	std::sort(&qlist[firstFinishedEntry], &qlist[qlistCnt], sortQuestIdx);
+
+	bool twoBlocks = (firstFinishedEntry != 0) && (firstFinishedEntry < qlistCnt);
+
+	topY = 0;
+	act2finSpacing = (!twoBlocks) ? 0 : lineHeight / 2;
+
+	int overallMinHeight = qlistCnt * lineHeight + act2finSpacing;
+	int space = panelInnerRect.size.height;
+
+	if (qlistCnt > 0) {
+		int additionalSpace = (space - overallMinHeight);
+		int addLineSpacing = additionalSpace / (qlistCnt);

```

Comment:
the parenthesis feels a bit inconsistent:

```suggestion
		int additionalSpace = space - overallMinHeight;
		int addLineSpacing = additionalSpace / qlistCnt;
```